File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    const visited = Array(dungeons.length).fill(false);
    let maxCount = 0; 
    function dfs(currentK, count) {
        maxCount = Math.max(maxCount, count);
        
        for (let i = 0; i < dungeons.length; i++) {
            const [need, use] = dungeons[i]
            if (!visited[i] && currentK >= need) {
                visited[i] = true;
                dfs(currentK - use, count + 1)
                visited[i] = false;
            }
        }
        
    }
    dfs (k, 0);
    return maxCount
    
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def backtracking(answer, k, count, visited, dungeons):
    # answer를 계속 갱신해주기
    answer = max(answer, count)
    n = len(dungeons)
    
    for i in range(n):
        limit, cost = dungeons[i]
        # 방문하지 않았고 k가 limit보다 크거나 같으면 count 처리
        if not visited[i] and limit <= k:
            visited[i] = True
            answer = backtracking(answer, k - cost, count + 1, visited, dungeons)
            visited[i] = False
    return answer
def solution(k, dungeons):
    answer = 0
    n = len(dungeons)
    visited = [False] * n
    count = 0
    
    answer = backtracking(answer, k, count, visited, dungeons)
    
    return answer